src/coreclr/jit/codegenxarch.cpp

```diff
@@ -6074,17 +6074,50 @@ void CodeGen::genCall(GenTreeCall* call)
     }
 #endif // defined(DEBUG) && defined(TARGET_X86)
 
-    // When it's a PInvoke call and the call type is USER function, we issue VZEROUPPER here
-    // if the function contains 256bit AVX instructions, this is to avoid AVX-256 to Legacy SSE
-    // transition penalty, assuming the user function contains legacy SSE instruction.
-    // To limit code size increase impact: we only issue VZEROUPPER before PInvoke call, not issue
-    // VZEROUPPER after PInvoke call because transition penalty from legacy SSE to AVX only happens
-    // when there's preceding 256-bit AVX to legacy SSE transition penalty.
-    // This applies to 512bit AVX512 instructions as well.
-    if (call->IsPInvoke() && (call->gtCallType == CT_USER_FUNC) && (GetEmitter()->Contains256bitOrMoreAVX()))
-    {
-        assert(compiler->canUseVexEncoding());
-        instGen(INS_vzeroupper);
+    var_types returnType = call->TypeGet();
+
+    if (call->IsPInvoke() && compiler->canUseVexEncoding())
```

Comment:
Hmmm. This works and is actually largely positive across the entirety of code.

However, it ends up pessimizing the scenario where you have a function that doesn't use any 256+ AVX and contains multiple P/Invokes.

It'd be better if we basically said "if the method does not use 256+ SIMD and method contains P/Invoke, emit `vzeroupper once in the prologue" and otherwise "emit vzeroupper before the P/Invoke" (with potential for a future optimization to only emit if there was some 256+ SIMD between the last vzeroupper and the current)

Reply:
Prototyped a change that should allow a basic version of this, using the same general tracking concept as we use for `Contains256bitOrMoreAVX`